tools/fp20compiler/ts1.0_inst_list.cpp

```diff
@@ -219,6 +219,61 @@ void InstList::Invoke()
     printf(");\n");
     printf("p += 2;\n");
 
+    // Process texture stage mode arguments
+    for (i=0; i<size; i++) {
+        switch(list[i].opcode.word) {
+        case TSP_OFFSET_2D_SCALE: {
+            float matrix[2*2] = {
+                list[i].args[1], list[i].args[2],
+                list[i].args[3], list[i].args[4]
+            };
+            float offset = list[i].args[5];
+            float scale = list[i].args[6];
+
+            printf("pb_push(p++, NV097_SET_TEXTURE_SET_BUMP_ENV_MAT + %d * 64, 6);\n", i);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[0] */\n", *(unsigned int*)&matrix[0]);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[1] */\n", *(unsigned int*)&matrix[1]);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[2] */\n", *(unsigned int*)&matrix[2]);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[3] */\n", *(unsigned int*)&matrix[3]);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_SCALE */\n", *(unsigned int*)&scale);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_OFFSET */\n", *(unsigned int*)&offset);
+            assert(false); /* Untested */
+            break;
+        }
+        case TSP_OFFSET_2D: {
+            float matrix[2*2] = {
+                list[i].args[1], list[i].args[2],
+                list[i].args[3], list[i].args[4]
+            };
+
+            printf("/* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT */\n");
+            printf("pb_push(p++, NV097_SET_TEXTURE_SET_BUMP_ENV_MAT + %d * 64, 4);\n", i);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[0] */\n", *(unsigned int*)&matrix[0]);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[1] */\n", *(unsigned int*)&matrix[1]);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[2] */\n", *(unsigned int*)&matrix[2]);
+            printf("*p++ = 0x%08x; /* NV097_SET_TEXTURE_SET_BUMP_ENV_MAT m[3] */\n", *(unsigned int*)&matrix[3]);
+            break;

```

Comment:
Did you try scaling horizontally / vertically separately, or rotations, to ensure the proper order of matrix elements?

Reply:
Yep. I just tested scaling horizontally and the matrix elements seem to be ordered fine.